### ISSUE TITLE
Refactor: Global app "back button" navigation

### ISSUE THREAD

**Issue by kyranjamie** [blockstack/blockstack-app#5181]
[Related discussion →](https://github.com/leather-wallet/extension/pull/5177#discussion_r1556928866)

Current solution is problematic in that `Container` component is home to all global app "What does back button do" logic. This is quickly growing in to a huge "do everything" function.

We should find a solution that colocates the back behaviour to the feature that has the back button (even if the button itself is in a generic header component). As a developer, if I need to modify back button logic for swap feature, I expect to see that logic in the swap feature.

One idea is to [decouple the back events using the PubSub pattern](https://github.com/leather-wallet/extension/pull/5177#discussion_r1557118103), but any solution that can move this logic outside of Container would be an improvement.

Suggest @pete-watters or @fbwoolf for this

**Comment by pete-watters** [comment]
I've been working on improving this as part of an overhaul of containers in https://github.com/leather-io/extension/pull/5619. 

A solution I came up with there is to wrap pages in `context` - `PageHeaderContext`  and then use that to update the `onBackLocation`. 

Pages that need this custom location are ones where we have the combination of:
- nested routed dialogs
- a back `<-` navigation in the header

Without providing this custom back location in these instances it can result in a loop of re-opening the routed dialog. 

For example in Swap:
- Go to `swap/STX/` 
- Choose an asset which opens the dialog and goes to `swap/STX/select-base`
- Close the dialog

If in this case we use `navigate(-1)` it will re-open the dialog as that was the previous page opened. We could use `navigate(-2)` but we still need to dynamically tell the header about that so instead I have updated the page header context to have an `onBackLocation` of `RouteUrls.Home`

```
useUpdatePageHeaderContext({ title: 'Swap', onBackLocation: RouteUrls.Home });
```
